### PR TITLE
New strategy for pytest - separating unit and integration tests

### PR DESCRIPTION
**Pull request by RaulRPrado**
Currently the tests run by pytest are strongly dependent on the DB connection and sim_telarray installation (which makes it hard to add it to a CI workflow). Ideally we should have "unit tests" independent of DB and simtel and "integration tests" specifics to each of them.

This PR presents a possible solution for this issue.

The general idea here is to mark the tests that require DB connection or sim_telarray installation and skip them, in case these are not available. This means that unit and integration tests will not be explicitly separated in the test modules, but they will be called separately depending on the availability of the DB and sim_telarray.

Why not separate all the test modules into unit and integration tests? First, this would mean rewriting all the tests modules and redesign them. Secondly, given the nature of our package (strong dependence with DB and sim_telarray), this separation can be quite difficult.

Now the technicalities.

Marking the tests in order to skip them is straightforward when using the the pytest decorator @pytest.mark.skipif. The more complicated step for us is to automatically identify whether the DB connection and sim_telarray installation are available.
I decided to do that by setting up a conftest.py file, that is loaded in the beginning of any pytest session. In this file I used the simtools config module to flag three things: a) the existence of the config file, b) the availability of the DB connection and c) the sim_telarray installation. 

a) is very simple. b) is also simple because we have a config entry called useMongoDB.
For c), I am looking at the sim_telarray path given in the config file and checking for the existence of the sim_telarray binary file.

The trick part was to pass these flags to the test modules. I tried to used pytest fixtures but it is not designed for that.
In the end I found a solution by using environment variables that are set in conftest.py and read by the modules. To encapsulate the reading of these variables I wrote some function in utils/tests.py.

Everything is working fine in the end. I have changed test_telescope_model and test_ray_tracing to exemplify how the test modules should me modified.

The next step is to add the skipif decorator to all test modules and also redesign some test modules in order to remove the dependence with the DB and simtel.

In summary, this is how our test setup will look like if this solution is accepted:

- The test modules run by pytest will contain unit tests and integration tests (DB and sim_telarray).
- If pytest is run in a environment without DB connection and sim_telarray installation (e.g. a CI environment), only the unit tests will be performed.
- Note that DB connection and sim_telarray are independent, so an environment may have only one of them available and only the respective integration tests will be run.

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    os.environ["HAS_CONFIG_FILE"] = "1"
+
+    # Checking whether sim_telarray is properly installed
+    simtelPath = cfg.get("simtelPath")
+    simtelBinPath = simtelPath + "/sim_telarray/bin/sim_telarray"
+    os.environ["SIMTEL_INSTALLED"] = "1" if os.path.exists(simtelBinPath) else "0"
+
+    # Checking whether there is DB connection
+    useMongoDB = cfg.get("useMongoDB")
```
Comment (orelgueta) [inline conftest.py:22]: I think this is not a good enough test. I think we should actually test if we can read from the DB instead. Maybe we can use one of the simple applications to get a value of a parameters (like one of the unit tests)? If we put it in a try/except block it should do the trick, no?

File: data/test-data/CTA-North-LST-1-Current_test-telescope-model.cfg [hunk 2]
```diff
@@ -0,0 +1,107 @@
+%==================================================
```
Comment (orelgueta) [inline data/test-data/CTA-North-LST-1-Current_test-telescope-model.cfg:1]: Is this what you referred to as "mock test data"? Just wondering what it is.

File: simtools/util/tests.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import os
+
+
+CONFIG_FILE_MSG = "Configuration file is not available"
```
Comment (orelgueta) [inline simtools/util/tests.py:4]: Change to "The gammasim-tool configuration file, config.yml, is not available" to be cleater.

File: simtools/util/tests.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+DB_CONNECTION_MSG = "There is not connection with DB"
```
Comment (orelgueta) [inline simtools/util/tests.py:6]: Change to "Connection with the DB is not available".

File: simtools/util/tests.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+SIMTEL_MSG = "sim_telarray is not installed"
+
+
+def _collect_conftest_flag(name):
```
Comment (orelgueta) [inline simtools/util/tests.py:11]: I suggest to write explicitly "config_test" instead of "conftest" everywhere. Also in the name of "conftest.py".

**Review by orelgueta — APPROVED** [review]
The idea looks good to me. As I mentioned, I would change the test of the DB to be explicit and a few other small stuff, but the concept is good.

File: conftest.py (hunk 1 above)
Comment (RaulRPrado) [reply conftest.py:22]: Isn't just creating a DatabaseHandler instance enough? It should crash if there is no DB connection, right?

Comment (RaulRPrado) [reply conftest.py:22]: No, it is not enough. We can create it successfully even with a broken connection.

Comment (RaulRPrado) [reply conftest.py:22]: I committed some changes. Now I use db.getModeParameters("north", "lst-1", "Current") to test whether the DB connection is fine.

File: data/test-data/CTA-North-LST-1-Current_test-telescope-model.cfg (hunk 2 above)
Comment (RaulRPrado) [reply data/test-data/CTA-North-LST-1-Current_test-telescope-model.cfg:1]: Not exactly. In this case I am just using an existing config file to create a telescope_model, instead of reading from the DB. The mock data framework provided by pytest is fancier than that. But I don't know much details.

File: simtools/util/tests.py (hunk 3 above)
Comment (RaulRPrado) [reply simtools/util/tests.py:4]: Done

File: simtools/util/tests.py (hunk 4 above)
Comment (RaulRPrado) [reply simtools/util/tests.py]: Done

File: simtools/util/tests.py (hunk 5 above)
Comment (RaulRPrado) [reply simtools/util/tests.py:11]: Not possible, the name conftest.py is hardcoded into pytest. Since this has to be the name of the file, I would keep in other places for consistency.

**Comment by RaulRPrado** [comment]
Thanks @orelgueta.
Please have a look at the changes and replies before I merge it.

I will merge this one and then work on a next PR adding the skipif to all the other test modules.

File: simtools/util/tests.py (hunk 5 above)
Comment (orelgueta) [reply simtools/util/tests.py:11]: really? how weird. Makes sense to keep consistent then and keep it everywhere.

File: conftest.py (hunk 1 above)
Comment (orelgueta) [reply conftest.py]: OK, looks good. Need to test creating the DatabaseHandler in various environments probably, but we can fix it later if we see issues.

**Comment by orelgueta** [comment]
It looks good, you can merge.
